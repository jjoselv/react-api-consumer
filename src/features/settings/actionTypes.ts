export const SET_NATIONALITY = 'SET_NATIONALITY';

export interface SetNationalityAction {
    type: typeof SET_NATIONALITY;
    payload: string;
};
    
export type SETTINGS_ACTION_TYPE = SetNationalityAction;
