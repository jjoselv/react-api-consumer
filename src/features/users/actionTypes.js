export const GET_USERS = 'GET_USERS';
export const GET_NEXT_BATCH = 'GET_NEXT_BATCH';
