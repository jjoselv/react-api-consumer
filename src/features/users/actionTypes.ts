import { User } from "../../components/Users/Users";

export const GET_USERS = 'GET_USERS';
export const GET_NEXT_BATCH = 'GET_NEXT_BATCH';
export type PromiseActions = 'PENDING' | 'FULFILLED' | 'REJECTED';

export interface SetUsersAction {
    type: `${typeof GET_USERS}_${PromiseActions}`;
    payload: {
        data: {
            results: User[];
            info: {
                page: number;
            };
        };
    };
};

export interface SetUsersNextBatchAction {
    type: `${typeof GET_NEXT_BATCH}_${PromiseActions}`;
    payload: {
        data: {
            results: User[];
            info: {
                page: number;
            };
        };
    };
};
    
export type USERS_ACTION_TYPE = SetUsersAction | SetUsersNextBatchAction;
