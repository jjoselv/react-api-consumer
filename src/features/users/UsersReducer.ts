import {User} from '../../components/Users/Users';
import {GET_NEXT_BATCH, GET_USERS, USERS_ACTION_TYPE} from './actionTypes';

export interface UsersState {
  users: undefined | User[],
  isLoading: boolean,
  hasError: boolean,
  isFulfilled: boolean,
  page: number,
}

const initialState: UsersState = {
  users: undefined,
  isLoading: false,
  hasError: false,
  isFulfilled: false,
  page: 1,
};

function reducer(state = initialState, action: USERS_ACTION_TYPE): UsersState {
  switch (action.type) {
    case `${GET_USERS}_PENDING`:
      return {
        ...state,
        isFulfilled: false,
        isLoading: true,
        hasError: false,
        users: undefined,
      };
    case `${GET_USERS}_FULFILLED`:
      return {
        ...state,
        isFulfilled: true,
        isLoading: false,
        hasError: false,
        users: action.payload.data?.results,
        page: action.payload.data?.info?.page,
      };
    case `${GET_USERS}_REJECTED`:
      return {
        ...state,
        isFulfilled: false,
        isLoading: false,
        hasError: true,
        users: undefined,
      };
    case `${GET_NEXT_BATCH}_PENDING`:
      return {
        ...state,
        isFulfilled: false,
        isLoading: true,
        hasError: false,
      };
    case `${GET_NEXT_BATCH}_FULFILLED`:
      return {
        ...state,
        isFulfilled: true,
        isLoading: false,
        hasError: false,
        users: [...(state.users ?? []), ...action.payload.data.results],
        page: action.payload.data?.info?.page,
      };
    case `${GET_NEXT_BATCH}_REJECTED`:
      return {
        ...state,
        isFulfilled: false,
        isLoading: false,
        hasError: true,
      };
    default:
      return state;
  }
};

export default reducer;
