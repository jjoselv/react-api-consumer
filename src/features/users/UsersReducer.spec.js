import {GET_USERS} from './actionTypes';
import UsersReducer from './UsersReducer';

describe('features > users > UsersReducer', () => {
  it('returns initial state, if non matched action is dispatched', () => {
    const initialState = {
      isLoading: false,
      hasError: false,
      isFulfilled: false,
    };

    const action = {
      type: 'FOO',
    };

    expect(UsersReducer(initialState, action)).toBe(initialState);
  });

  /**
   * Provide table of values to run test case against
   * @see https://jestjs.io/docs/en/api#testeachtablename-fn-timeout
   */
  it.each([
    [`${GET_USERS}_FULFILLED`],
    [`${GET_USERS}_PENDING`],
    [`${GET_USERS}_REJECTED`],
  ])(`updates state according to dispatched action`, actionType => {
    const initialState = {
      users: undefined,
      isLoading: false,
      hasError: false,
      isFulfilled: false,
      page: 1,
    };

    const payload =
      actionType === `${GET_USERS}_FULFILLED`
        ? {
            data: {
              results: [{a: 2}],
              info: {
                page: 42,
              },
            },
          }
        : undefined;

    const action = {
      type: actionType,
      payload,
    };

    expect(UsersReducer(initialState, action)).toMatchSnapshot();
  });
});
