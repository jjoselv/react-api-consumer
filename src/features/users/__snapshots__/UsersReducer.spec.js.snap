// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features > users > UsersReducer updates state according to dispatched action 1`] = `
Object {
  "hasError": false,
  "isFulfilled": true,
  "isLoading": false,
  "page": 42,
  "users": Array [
    Object {
      "a": 2,
    },
  ],
}
`;

exports[`features > users > UsersReducer updates state according to dispatched action 2`] = `
Object {
  "hasError": false,
  "isFulfilled": false,
  "isLoading": true,
  "page": 1,
  "users": undefined,
}
`;

exports[`features > users > UsersReducer updates state according to dispatched action 3`] = `
Object {
  "hasError": true,
  "isFulfilled": false,
  "isLoading": false,
  "page": 1,
  "users": undefined,
}
`;
